### PR TITLE
DB API 2.0

### PR DESCRIPTION
**Pull request by barbieri**

### CODE REVIEW

**Comment by barbieri** [comment]
This PR should wait on #7 as most of the commits will be reviewed there.

**Comment by honzasp** [comment]
I will have a closer look after we deal with #7 :)

**Comment by barbieri** [comment]
@honzasp, it's now ready to be reviewed. For the API docs, I used sphinx autodoc, so it will get the function/class docstrings.

The test suite only runs on python 3.11 since I used its code, some stuff it check is not available in older version. You can see changes to their code by downloading and diff cpython-3.11.3/Lib/test/test_sqlite3

File: libsql_client/dbapi2/types.py [hunk 1]
```diff
@@ -0,0 +1,1423 @@
...
+        # https://github.com/python/cpython/blob/main/Modules/_sqlite/statement.c#L32
+        # however we need to parse the statement with iter_sql_statements()
+        # as we can't use sqlite3_prepare_v2()
+        itr = iter(iter_sql_statements(sql))
+        self.tokens = next(itr)
+        assert self.tokens
+        next_stmt = next(itr, None)
+        if next_stmt and next_stmt[0].upper() != "END":
+            msg = "You can only execute one statement at a time."
+            raise ProgrammingError(msg)
```
Comment (honzasp) [inline libsql_client/dbapi2/types.py:239]: Why we ignore this error when the second statement is `END`?

File: libsql_client/dbapi2/types.py [hunk 2]
```diff
@@ -0,0 +1,1423 @@
...
+        # https://www.sqlite.org/lang_delete.html
+        # Syntax:
+        #    WITH RECURSIVE table_name ( col_name, other_col)
+        #    AS NOT MATERIALIZED ( SELECT ... )
+        return tokens[0].upper() in cls._dml_statements
+
+    @property
+    def is_readonly(self) -> bool:
+        # mimics sqlite3_stmt_readonly()
+        return not self.is_dml
```
Comment (honzasp) [inline libsql_client/dbapi2/types.py:258]: This does not mimic `sqlite3_stmt_readonly()`: for example, `CREATE TABLE` is not a readonly statement.

File: libsql_client/dbapi2/types.py [hunk 3]
```diff
@@ -0,0 +1,1423 @@
...
+
+    :meta private:
+    """
+    def wrapper(*args: P.args, **kwargs: P.kwargs) -> T:
+        self = args[0]
+        assert len(args) > 0 and isinstance(self, Connection)
+        self._do_check_connection()
+        return fn(*args, **kwargs)
+
+    return wrapper
```
Comment (honzasp) [inline libsql_client/dbapi2/types.py:290]: What's the point of using a decorator instead of simply calling the `_check` method?

File: libsql_client/dbapi2/types.py [hunk 4]
```diff
@@ -0,0 +1,1423 @@
...
+
+    def __init__(
+        self,
+        database: str,
+        timeout: float,
+        detect_types: int,
+        isolation_level: Optional[IsolationLevel],
+        check_same_thread: bool,
+        cached_statements: int,
+        autocommit: Autocommit,
```
Comment (honzasp) [inline libsql_client/dbapi2/types.py:323]: What is this `autocommit` parameter? I can't find it in the original sqlite3.

**Review by honzasp — COMMENTED** [review]
I did a first pass through the code, but I need a break before continuing :D

**Review by honzasp — COMMENTED** [review]
I did a first pass through the code, but I need a break before continuing :D

**Comment by barbieri** [comment]
I didn't run any benchmarks, but I don't see why it would matter, unless you see some clear optimization to be employed. The synchronous characteristics will already add lots of latency

File: examples/dbapi2.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
+from libsql_client import dbapi2
+import datetime
+import os
+import logging
+
+
+def output(*text):
+    print(*text)
```
[the remark this answers is not in the document]
Comment (barbieri) [reply examples/dbapi2.py:8]: ok, not a big deal... I changed this in my last commits to make it easier to read the rest of the code. But I can go back to plain print().

Also, compared to the other example, this is a more complete showcase, we used it as a test suite before the actual tests were added... if you think it's too big, we can shrink it as well.

File: libsql_client/dbapi2/_async_executor.py [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+
+if sys.version_info[:2] >= (3, 9):
+    MainQueue = queue.Queue[LoopControl]
+else:
+    MainQueue = queue.Queue
+
+
+# NOTE: keep outside of class, do not touch self, not even keep a reference
+def _thread_main(log_prefix: str, q: MainQueue) -> None:
+    loops: List[asyncio.AbstractEventLoop] = []
```
[the remark this answers is not in the document]
Comment (barbieri) [reply libsql_client/dbapi2/_async_executor.py:43]: with python this is the standard way to get a pointer. wanted to have `loops` set/used from within the function.

There is a keyword `nonlocal`, but I rarely use or see it in use...

File: libsql_client/dbapi2/_async_executor.py [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+            except Exception as e:
+                self._dbg(
+                    "failed: %s, args=%s, kwargs=%s, exc=%s",
+                    fn, args, kwargs, e,
+                    exc_info=e,
+                )
+                raise
+
+        coro = run_in_main_loop()
+        return asyncio.run_coroutine_threadsafe(coro, self._control.loop)
```
[the remark this answers is not in the document]
Comment (barbieri) [reply libsql_client/dbapi2/_async_executor.py:213]: this is done by asyncio.run_coroutine_threadshafe, it will add a future and handle the cases when it should fail.

However, given the synchronous usage and the rest of the code, this can't happen, the connection close will send the termination and dispose the object right away

File: libsql_client/dbapi2/_utils.py [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+
+def log_prefix(
+        logger: logging.Logger,
+        prefix: str,
+        level: int,
+        msg: str,
+        *args: object,
+        exc_info: Optional[BaseException] = None,
+) -> None:
+    logger.log(level, prefix + msg, *args, exc_info=exc_info)
```
[the remark this answers is not in the document]
Comment (barbieri) [reply libsql_client/dbapi2/_utils.py:14]: I like it, since with `log_obj` + `functools.partial` we can get properly crafted messages without typing too much.

File: libsql_client/dbapi2/_utils.py [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+            # NOTE: in C pos[end_pos] == 0, so "pos[1] == '-'" is enough,
+            # Here we must compare pos + 1 < end_pos
+            if pos + 1 < end_pos and sql[pos + 1] == "-":
+                pos += 2
+                # NOTE: in C pos[end_pos] == 0, so they do "pos[0]".
+                # Here we compare pos + 1 < end_pos
+                while pos < end_pos and sql[pos] != "\n":
+                    pos += 1
+                if pos >= end_pos:
+                    return None
```
[the remark this answers is not in the document]
Comment (barbieri) [reply libsql_client/dbapi2/_utils.py:58]: this is a port of the C function, they do that 🤷

**Comment by honzasp** [comment]
I definitely see many applications where latency matters. If it turns out that code that switches from sqlite3 to Turso becomes 1000x slower, we will have hard time selling this to people :)

File: libsql_client/dbapi2/_utils.py [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+
+                pos += 1  # C uses for() with trailing ++
+                continue
+
+            return sql[pos:]  # not a line comment
```
[the remark this answers is not in the document]
Comment (barbieri) [reply libsql_client/dbapi2/_utils.py:63]: this is a port 1:1 to the C function, with matching comments.

I want to keep as close as possible to behave the same. There are much easier ways to implement this with Python, but in this project whenever possible the code is the same as in their C.

If there are different behaviors, such as in the C changed, we should change it here to match.

File: libsql_client/dbapi2/_utils.py [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+
+
+_iter_sql_delim_chars = {";", ",", "(", ")", "[", "]"}
+_iter_sql_stop_chars = _lstrip_sql_whitespace_chars.union(
+    _iter_sql_delim_chars,
+)
+_iter_sql_quote_chars = {"\"", "'"}
+
+
+def _iter_sql_get_quoted_end(sql: str, pos: int, end_pos: int) -> int:
```
[the remark this answers is not in the document]
Comment (barbieri) [reply libsql_client/dbapi2/_utils.py:101]: because I had the value before, not a big deal

File: libsql_client/dbapi2/_utils.py [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+    pos += 1
+    while pos < end_pos:
+        if escaped:
+            escaped = False
+        else:
+            if sql[pos] == ch:
+                pos += 1
+                break
+            escaped = sql[pos] == "\\"
+        pos += 1
```
[the remark this answers is not in the document]
Comment (barbieri) [reply libsql_client/dbapi2/_utils.py:116]: hum, that's true. I'll fix

File: libsql_client/dbapi2/_utils.py [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+                pos += 1
+
+        if pos == end_pos:
+            yield sql
+            break
+
+        if pos > 0:
+            yield sql[:pos]
+            sql = sql[pos:]
+        elif sql[0] in _iter_sql_delim_chars:
```
[the remark this answers is not in the document]
Comment (barbieri) [reply libsql_client/dbapi2/_utils.py:177]: it's (or should be) impossible to happen

File: libsql_client/dbapi2/_utils.py [hunk 14]
```diff
@@ -0,0 +1,199 @@
...
+                pos = _iter_sql_get_quoted_end(sql, pos, end_pos)
+                break
+            elif ch in _iter_sql_stop_chars:
+                break
+            else:
+                pos += 1
+
+        if pos == end_pos:
+            yield sql
+            break
```
[the remark this answers is not in the document]
Comment (barbieri) [reply libsql_client/dbapi2/_utils.py:172]: that's true

File: libsql_client/dbapi2/types.py [hunk 15]
```diff
@@ -0,0 +1,1423 @@
...
+            try:
+                self.rollback()
+            except Exception as e:  # TODO: is this explicitly needed?
+                raise e from exc_value
+        return False
+
+    @check_thread
+    @check_connection
+    def __call__(self, sql: str) -> Statement:
+        return Statement(sql, _allow_instantiation=self)
```
[the remark this answers is not in the document]
Comment (barbieri) [reply libsql_client/dbapi2/types.py:411]: yeah, I don't get the reason for it either... but they do have that and they do test that. :-(

File: libsql_client/dbapi2/hrana.py [hunk 16]
```diff
@@ -0,0 +1,423 @@
...
+        error: Optional[BaseException],
+) -> RawExecuteResult:
+    if isinstance(error, aiohttp.ClientError):
+        code, name = _get_aiohttp_client_error_code(error)
+        error = OperationalError(str(error))
+        if code:
+            error.sqlite_errorcode = code  # type: ignore
+            error.sqlite_errorname = name  # type: ignore
+
+    return RawExecuteResult([result], [error])
```
[the remark this answers is not in the document]
Comment (barbieri) [reply libsql_client/dbapi2/hrana.py:161]: the rest of the code handles batch (executemany) and single execute as the same.

File: libsql_client/dbapi2/hrana.py [hunk 17]
```diff
@@ -0,0 +1,423 @@
...
+    @overload
+    def run_in_executor(fn: Callable[P, Awaitable[T]]) -> Callable[P, T]:
+        ...
+
+    @overload
+    def run_in_executor(fn: Callable[P, T]) -> Callable[P, T]:
+        ...
+
+
+def run_in_executor(fn: Callable[P, T]) -> Callable[P, T]:
```
[the remark this answers is not in the document]
Comment (barbieri) [reply libsql_client/dbapi2/hrana.py:196]: this is the actual implementation, it handles the above overloads... so it doesn't really matter the signature, since T may be modified.

File: libsql_client/dbapi2/hrana.py [hunk 18]
```diff
@@ -0,0 +1,423 @@
...
+    :py:class:`AsyncExecutor` thread by doing a ``AsyncExecutor.submit()``
+    and then ``future.result(timeout)``.
+
+    The method itself will block until the executor runs.
+
+    :meta private:
+    """
+    def wrapper(*args: P.args, **kwargs: P.kwargs) -> T:
+        self = args[0]
+        assert len(args) > 0 and isinstance(self, ConnectionHrana)
```
[the remark this answers is not in the document]
Comment (barbieri) [reply libsql_client/dbapi2/hrana.py:209]: good catch

File: libsql_client/dbapi2/hrana.py [hunk 19]
```diff
@@ -0,0 +1,423 @@
...
+
+    def _raw_execute(
+            self,
+            sql: str,
+            parameters: Iterable[SqlParameters],
+            *,
+            want_rows: bool = True,
+    ) -> RawExecuteResult:
+        stmts = _conv_stmts(sql, parameters, want_rows)
+        assert len(stmts) > 0, "Expected at least 1 parameter"
```
[the remark this answers is not in the document]
Comment (barbieri) [reply libsql_client/dbapi2/hrana.py:404]: because it can never happen. I noticed this bug and it happened because I mistyped the default argument, instead of `((),)` it was `()`.

decided to leave this assertion, note that running python with `-OO` will remove assertions.

File: libsql_client/dbapi2/types.py [hunk 20]
```diff
@@ -0,0 +1,1423 @@
...
+
+class RawExecuteResult:
+    """Iterable over pairs of ``result, error``.
+
+    :meta private:
+    """
+    results: List[Optional[RawResults]]
+    errors: List[Optional[BaseException]]
+    _idx: int
+    __slots__ = ("results", "errors", "_idx")
```
[the remark this answers is not in the document]
Comment (barbieri) [reply libsql_client/dbapi2/types.py:158]: because you return two lists in your API, this class makes it easy to use that without bothering.

Although we have a `hrana.py` and `types.py`, since all will be hrana, it's nore like `websockets.py` and `types.py` composed of hrana stuff. But I won't rename these until we come to implement the http... if websockets is gone, then I'll just replace the contents of hrana.py, if we keep both, I'll likely rename the file.

File: libsql_client/dbapi2/types.py (hunk 1 above)
Comment (barbieri) [reply libsql_client/dbapi2/types.py:239]: they have a test with `CREATE TRIGGER ... BEGIN ... ; END`. From their C code, this is handled as a single thing by https://www.sqlite.org/c3ref/prepare.html

that's what they call "a statement". They don't do the parsing themselves... I can't call C to check, so that's the best I can do to keep the same behavior.

File: libsql_client/dbapi2/types.py (hunk 2 above)
Comment (barbieri) [reply libsql_client/dbapi2/types.py:258]: do you have more that I should be handling? Or just "look at sqlite c code"?

but ok, needs fixing

Comment (barbieri) [reply libsql_client/dbapi2/types.py:258]: added ddl: "CREATE", "ALTER", "DROP", let me know if more is needed.

File: libsql_client/dbapi2/types.py (hunk 4 above)
Comment (barbieri) [reply libsql_client/dbapi2/types.py:331]: yeah, not documented but exists in their code, is checked by their test suite and greatly impacts the behavior :-/

File: libsql_client/dbapi2/types.py [hunk 21]
```diff
@@ -0,0 +1,1423 @@
...
+        self.row_factory = None
+        self.text_factory = None
+        self.cursor_factory = Cursor
+        self._autocommit = autocommit
+        self._trace_callback = None
+        self._limits = _default_limits.copy()
+        self.isolation_level = isolation_level
+        self._raw_init()  # after this point we can call the DB
+        if autocommit is False:
+            self.cursor()._query_execute("BEGIN", want_rows=False)
```
[the remark this answers is not in the document]
Comment (barbieri) [reply libsql_client/dbapi2/types.py:354]: it will skip some of the autocommit logic, it impacts when you do a commit/rollback, executescript.

It's following exactly what they do, so you may refer to their C python bindings

File: examples/dbapi2.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (barbieri) [reply examples/dbapi2.py]: changed

File: libsql_client/dbapi2/_async_executor.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (barbieri) [reply libsql_client/dbapi2/_async_executor.py]: changed

File: libsql_client/dbapi2/_utils.py [hunk 22]
```diff
@@ -0,0 +1,199 @@
...
+        if not sql:
+            break
+
+        pos = 0
+        end_pos = len(sql)
+        while pos < end_pos:
+            ch = sql[pos]
+            if ch in _iter_sql_quote_chars:
+                pos = _iter_sql_get_quoted_end(sql, pos, end_pos)
+                break
```
[the remark this answers is not in the document]
Comment (honzasp) [reply libsql_client/dbapi2/_utils.py:189]: We are also changing behavior depending on whether the statement is an `UPDATE`/`DELETE`/`INSERT` or something else. `UPDATE"t" SET x = 0` would be parsed as non-UPDATE by this code, but as UPDATE by SQLite.

File: libsql_client/dbapi2/_utils.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (honzasp) [reply libsql_client/dbapi2/_utils.py:198]: Are you OK with falling into an infinite loop if this invariant breaks in the future? This code is very hard to follow, it will be very easy for future changes to break it.

File: libsql_client/dbapi2/hrana.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (honzasp) [reply libsql_client/dbapi2/hrana.py:404]: So calling `cursor.executemany("SELECT ?", [])` is forbidden?

File: libsql_client/dbapi2/types.py (hunk 3 above)
Comment (honzasp) [reply libsql_client/dbapi2/types.py:298]: I don't understand how

```
def foo(self):
    self._check_connection()
    ...
```

is harder to read and less polluting than

```
@check_connection
def foo(self):
    ...
```

But if it's just a matter of personal preference and there is no functional difference, then why not; you are going to maintain this thing, after all :)

File: libsql_client/dbapi2/_utils.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (barbieri) [reply libsql_client/dbapi2/_utils.py:201]: I can do an assert in the else clause... but it can never happen since the only way to stop at the first char (pos = 0) is when we face a _iter_sql_delim_chars

[the remark this answers is not in the document]
Comment (barbieri) [reply libsql_client/dbapi2/_utils.py]: anyway, changed to an `else`, as it should always be the case (elif condition must always be true)

File: libsql_client/dbapi2/hrana.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (barbieri) [reply libsql_client/dbapi2/hrana.py]: fixed

File: libsql_client/dbapi2/hrana.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (barbieri) [reply libsql_client/dbapi2/hrana.py]: while `executemany()` doesn't allow SELECT, it does allow such empty list for insert... fixed.

however I also noticed that the behavior cannot be 100% matching since sqlite3 will prepare the statement even if no parameters are given... then you can get exceptions if they are invalid.

File: libsql_client/dbapi2/types.py (hunk 3 above)
Comment (barbieri) [reply libsql_client/dbapi2/types.py:299]: the decorator could also change docstrings and the likes, it's a more common pattern

File: libsql_client/dbapi2/_async_executor.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (barbieri) [reply libsql_client/dbapi2/_async_executor.py:213]: not sure I get what you mean, you mean 2 threads calling close() at the same time? If so I tested and you get `RuntimeError: Event loop is closed`.

I don't think it's worth to add locking around `self._control`, but could do that if you wish

File: libsql_client/dbapi2/hrana.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (honzasp) [reply libsql_client/dbapi2/hrana.py]: You can ensure that the statement is prepared (and exceptions are raised) by sending a `describe` Hrana request if the list is empty, but that's probably not necessary.

File: libsql_client/dbapi2/types.py (hunk 1 above)
Comment (honzasp) [reply libsql_client/dbapi2/types.py:243]: The `CREATE TRIGGER` statement can contain multiple statements before the `END`, i.e. `CREATE TRIGGER ... BEGIN SELECT 1; SELECT 2; SELECT 3; END`.
https://www.sqlite.org/lang_createtrigger.html

I think that we should either not support `CREATE TRIGGER` at all, or try to parse it correctly by checking for statements that begin with tokens `CREATE`, `TRIGGER` and looking for the matching `END`. Supporting only CREATE TRIGGERs that contain exactly one statement will be confusing.

File: libsql_client/dbapi2/types.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (honzasp) [reply libsql_client/dbapi2/types.py:355]: Let's hope that people won't use this mode in practice, it won't work well with Turso :)

File: libsql_client/dbapi2/types.py [hunk 23]
```diff
@@ -0,0 +1,1432 @@
...
+        # _pysqlite_query_execute()
+        # https://github.com/python/cpython/blob/main/Modules/_sqlite/cursor.c#L867-L877
+        if self._autocommit is LEGACY_TRANSACTION_CONTROL \
+           and self.isolation_level is not None \
+           and statement.is_dml \
+           and self._sqlite3_get_autocommit():
+            self._begin_transaction()
+
+    _begin_transaction_tokens: ClassVar[Set[str]] = {"BEGIN"}
+    _end_transaction_tokens: ClassVar[Set[str]] = {"COMMIT", "ROLLBACK"}
```
[the remark this answers is not in the document]
Comment (barbieri) [reply libsql_client/dbapi2/types.py]: added, thanks!

**Comment by barbieri** [comment]
@honzasp squashed all the fixups, the PR is ready to be merged. Thanks for your help 🙏

File: libsql_client/dbapi2/types.py (hunk 1 above)
Comment (barbieri) [reply libsql_client/dbapi2/types.py:243]: yeah, I'm not even sure Turso supports those, I recall having to disable the test case for that and it wasn't related to my part. Do you know?

Alternatively I though of a hack to simplify this part, but it's kinda ugly: I can keep a singleton `connection(":memory:")` and call it (`__call__()`), as you see in my code, which matches the C, all it does is to call prepare on the statement and evaluate it...

File: libsql_client/dbapi2/types.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (barbieri) [reply libsql_client/dbapi2/types.py:355]: their test suite passes 🤷
